go/felix/ifacemonitor/iface_monitor.go

```diff
@@ -89,26 +106,100 @@ func (m *InterfaceMonitor) handleNetlinkUpdate(update netlink.LinkUpdate) {
 	}
 	msgType := update.Header.Type
 	ifaceExists := msgType == syscall.RTM_NEWLINK // Alternative is an RTM_DELLINK
-	m.storeUpdateAndNotifyOnChange(ifaceExists, attrs)
+	m.storeAndNotifyLink(ifaceExists, update.Link)
+}
+
+func (m *InterfaceMonitor) handleNetlinkAddrUpdate(update netlink.AddrUpdate) {
+	addr := update.LinkAddress.IP.String()
+	ifIndex := update.LinkIndex
+	exists := update.NewAddr
+	log.WithFields(log.Fields{
+		"addr":    addr,
+		"ifIndex": ifIndex,
+		"exists":  exists,
+	}).Info("Netlink address update.")
+
+	if _, known := m.ifaceName[ifIndex]; !known {
+		log.WithField("ifIndex", ifIndex).Warn("No known iface with this index.")
+		return
+	}
+	if _, known := m.ifaceAddrs[ifIndex]; !known {
+		// We think this interface is down, or haven't yet heard about it at all -
+		// indicates a race between the link and address update channels.
+		// Addresses will be notified when we process the link update.
+		log.WithField("ifIndex", ifIndex).Warn("Race for new interface.")
+		return
+	}
+
+	if exists {
+		if !m.ifaceAddrs[ifIndex].Contains(addr) {
+			m.ifaceAddrs[ifIndex].Add(addr)
+			m.notifyIfaceAddrs(ifIndex)
+		}
+	} else {
+		if m.ifaceAddrs[ifIndex].Contains(addr) {
+			m.ifaceAddrs[ifIndex].Discard(addr)
+			m.notifyIfaceAddrs(ifIndex)
+		}
+	}
+}
+
+func (m *InterfaceMonitor) notifyIfaceAddrs(ifIndex int) {
+	if m.ifaceAddrs[ifIndex] != nil {
+		m.AddrCallback(m.ifaceName[ifIndex], m.ifaceAddrs[ifIndex].Copy())
```

Comment:
Ah, sorry, no, `Set` is an interface and a nil interface is not the same as a nil pointer to a struct that implements an interface.  One of go's gotchas.